pkg/types/chain_reader.go

```diff
@@ -0,0 +1,60 @@
+package types
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type chainReaderError string
+
+func (e chainReaderError) Error() string { return string(e) }
+
+const (
+	ErrInvalidType   = chainReaderError("invalid type")
+	ErrFieldNotFound = chainReaderError("field not found")
+	ErrInvalidConfig = chainReaderError("invalid configuration")
+)
+
+func UnwrapClientError(err error) error {
+	if s, ok := status.FromError(err); ok {
+		if s.Code() == codes.Unimplemented {
+			return errors.ErrUnsupported
+		}
+		return chainReaderError(s.String())
+	}
+	return err
+}
+
+type ChainReader interface {
+	// returnVal should satisfy Marshaller interface
```

Comment:
Not a blocker, but shouldn't this have the main docs? I thought we a longer description for this, with examples.

Reply:
Yeah this comment was lost, and IIRC we had more details on the proposal document.

Reply:
Ah, good catch